woocommerce/woocommerce
# Introduce a simplified dependency injection container

### Changes proposed in this Pull Request:

This pull request introduces a simplified dependency injection container as a replacement of the PHP League's Container package. The motivation for this change is the performance gain, and the fact that we are effectively only using the container for one task: resolving single-instance classes by class name (we don't need a full-blown dependency injection engine for that).

This pull request introduces two new classes:

* `RuntimeContainer`: this class is a replacement for `ExtendedContainer` as the underlying dependency injection container used by WooCommerce at runtime. It doesn't implement explicit class registration, instead it considers any class belonging to the `Automattic\WooCommerce` namespace as implicitly registered as a single-instance class. It uses reflection to resolve (recursively) the dependencies passed as parameters to the `init` method of the resolved classes.
* `TestingContainer`: inheriting from the previous one, includes additional methods to replace the regular classes with mocks and it's intended for usage when running unit tests (being, too, a replacement for `ExtendedContainer`).

`RuntimeContainer` is aware of the containers that the Store API and Blocks implement on their own, and will delegate class resolutions to these containers for classes within the respective namespaces.

Temporarily it's possible to force WooCommerce to still use `ExtendedContainer` at runtime by either creating a `WOOCOMMERCE_USE_OLD_DI_CONTAINER` constant with a value of `true`, or hooking on the `woocommerce_use_old_di_container` filter and having it returning true. This should be done only if issues are found in the new container and the site is broken as a result. `ExtendedContainer` (along with the PHP Lague's container package and all the supporting code) will be removed in WooCommerce 10.0.

Smilarly, it's possible to force WooCommerce to still use `ExtendedContainer` for unit tests by defining the `USE_OLD_DI_CONTAINER` environment variable before triggering the tests run.

### How to test the changes in this Pull Request:

1. Verify that the WooCommerce critical flows (like placing an order from the store and managing products and orders from admin) work with and without adding `define('WOOCOMMERCE_USE_OLD_DI_CONTAINER', true);` in `wp-config.php` (or alternatively, with and without adding this snippet: `add_filter('woocommerce_use_old_di_container', '__return_true');`
2. Verify that all the unit tests pass.
3. Temporarily add the following method to the `Container` class (`src/Container.php`), then run `wp eval 'wc_get_container()->instantiate_everything();'`, again with both the old and the new containers (see step 1):

```php
// Use the current container to retrieve an instance of each of the classes that are registered with the old container.
public function instantiate_everything() {
    foreach ( $this->get_service_providers() as $provider ) {
        $refclass = new \ReflectionClass( $provider );
        $provides_prop = $refclass->getProperty( 'provides' );
        $provides_prop->setAccessible( true );
        $provides_list = $provides_prop->getValue( $refclass->newInstance() );
        foreach ( $provides_list as $class_name ) {
            echo get_class( $this->get( $class_name ) ) . "\n";
        }
    }
}
```

4. Review [the notice added to `src/README.md`](https://github.com/woocommerce/woocommerce/blob/simplify-dependency-injection-container/plugins/woocommerce/src/README.md#important-notice).

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
